tests/component-without-proptypes.d.ts

```diff
@@ -8,5 +8,5 @@ declare module 'component' {
     render(): JSX.Element;
   }
 
-  export function test(): JSX.Element;
+  export type test = React.SFC;

```

Comment:
Regarding the `///` directives question:

Given a component like this:

```ts
import * as React from 'react';

export const Test: React.SFC = () => <div />;
```

`tsc` will generate this `d.ts` file:

```ts
/// <reference types="react" />
import * as React from 'react';
export declare const Test: React.SFC;
```

This is kind of surprising since I would have expected either the `///` reference *or* the import. In fact, `tslint` will complain about this (`No need to reference "react", since it is imported. (no-reference-import)`). Removing either one of the first two lines still results in a valid declaration for the component. It kind of gets messy though when there is also a class component in the module. For this module:

```ts
import * as React from 'react';

export const Test: React.SFC = () => <div />;
export class Test2 extends React.Component {}
```

`tsc` generates:

```ts
/// <reference types="react" />
import * as React from 'react';
export declare const Test: React.SFC;
export declare class Test2 extends React.Component {
}
```

`react2dts` would generate this though:

```ts
import {Component} from 'react';
export const Test: React.SFC;
export class Test2 extends Component {}
```

And this is now invalid since `React` is not imported with a namespace import. (See also this test fixture I'm commenting on.)

Therefore I would suggest we either:

1. get rid of the `///` reference altogether and adjust the react import statement to always be a namespace import instead, or
2. always insert the `///` directive when there is an SFC part of the module. In this case the above example would become:

    ```ts
    /// <reference types="react" />
    import {Component} from 'react';
    export const Test: React.SFC;
    export class Test2 extends Component {
    }
    ```

The example in 2.) is valid but looks strange. I would prefer solution 1. What do you guys think, @KnisterPeter @ingdir @DeTeam?

Reply:
I would prefer 1. as well

Reply:
(2) would probably be easier to implement, but I agree (1) is cleaner. Can be added as part of this PR.

Reply:
Yes, it should be done in this PR.

Reply:
Pushed some changes, @ingdir @KingHenne @KnisterPeter 
Not sure if we have to do more clean up.

While reading realized that there might be some issues even with the current setup — made a failing test-case for it: https://github.com/KnisterPeter/react-to-typescript-definitions/pull/483